WIP: improve support for crosscompiling to musl32

wfranzini opened the pull request:
###### Motivation for this change

Crosscompiling packages to musl32 lead to compilation errors due to the missing `__stack_chk_fail_local` symbol.  Even building nix fails. This PR disable stackprotector for musl32.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

wfranzini committed:
disable stackprotector on target platform musl32

wfranzini committed:
libexecinfo: enables stackprotector through hardeningEnable

The makefile patch pulled from alpinelinux used to enable
stackprotector unconditionally.  This change use the nixpkgs flag
hardeningEnable.

This is required to support platform where stackprotector has
problems.

wfranzini committed:
libsodium: disable stackprotector on musl32

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/crosscompilation-to-musl32-problems/3110/5

wfranzini committed:
audit: backport from upstream support for system w/o strndupa/rawmemchr

The patches from alpinelinux failed to apply to v2.8.5, on the other
hand upstream has added support for musl based system in the master
branch on github [1].

The patch has been added to the nixpkgs repo since it needs to be
modified slightly, a copyright notice has been removed, to apply
cleanly.

1. https://github.com/linux-audit/audit-userspace/commit/d579a08bb1cde71f939c13ac6b2261052ae9f77e

flokli commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -343,7 +343,9 @@ stdenv.mkDerivation {
     '' + optionalString (targetPlatform.libc == "newlib") ''
       hardening_unsupported_flags+=" stackprotector fortify pie pic"
-    '' + optionalString targetPlatform.isNetBSD ''
+      '' + optionalString (targetPlatform.libc == "musl" && targetPlatform.isi686) ''
```
is this only a problem for i686, or 32 bit in general?

wfranzini replied:
musl32 is declared as i686-unknown-linux-musl

matthewbauer replied:
I got the same issue on i486 and i586 as well. This should be ```targetPlatform.isx86_32``` to cover those cases.

matthewbauer replied:
Make sure spacing is consistent here as well.

wfranzini committed:
bison2: add perl in nativeBuildInputs

Without this change crosscompiling fails because perl is not found
during configure.

wfranzini committed:
sox: musl does not support pipe rewind

Disable the pipe rewind related feature since musl has no support for it.

matthewbauer commented on pkgs/development/libraries/libexecinfo/default.nix:
```diff
@@ -20,14 +20,11 @@ stdenv.mkDerivation rec {
-      url = https://git.alpinelinux.org/cgit/aports/plain/main/libexecinfo/30-linux-makefile.patch?id=730cdcef6901750f4029d4c3b8639ce02ee3ead1;
-      sha256 = "1jwjz22z5cjy5h2bfghn62yl9ar8jiqhdvbwrcfavv17ihbhwcaf";
-    })
+    ./30-linux-makefile.patch
```
I wonder if alpine has a way to get ssp working?

wfranzini replied:
@matthewbauer Alpinelinux creates a small libssp_nonshared library in the musl-dev package.
https://git.alpinelinux.org/aports/tree/main/musl?h=master

I've recreated it in my local nixpkgs repo, however I'm not sure how to automatically link with libssp_nonshared.a.

Do you have a hint?

matthewbauer replied:
You probably need to put it in musl's $out/lib/libssp_nonshared.a. Unfortunately we also need to apply this patch to GCC:

https://git.alpinelinux.org/aports/tree/main/gcc/gcc-6.1-musl-libssp.patch

That would probably be okay to apply to GCC, but maybe only on i686-musl builds.

wfranzini committed:
identify musl32 using x86_32 instead of i686

7c6f434c commented:
For context: is this problem irrelevant in 64-bit case?

matthewbauer commented:
> For context: is this problem irrelevant in 64-bit case?

I have never seen it on x86_64. I think it could be possible to hit it with some messy linking, but I would prefer to limit the patching to just i*86 for now.

wfranzini committed:
add stackprotector support on musl32

In order to support stackprotector on musl32, this change import a
couple of patches from alpinelinux:
1. libssp_nonshared.a is built alongside musl's libc
2. the above library is automatically linked when compiling with gcc6
   or gcc7

symphorien commented:
I successfully used this PR to compile openssl.

Ericson2314 commented:
Can we submit some of these patches upstream and `fetchpatch` them?

wfranzini commented:
> Can we submit some of these patches upstream and `fetchpatch` them?

@Ericson2314  I'll look into this the next week.  

I would like to have this PR merged somewhere in the future, but I need guidance about the correct branch to use since I fear it will cause a massive rebuild.

flokli commented:
@wfranzini this should target the `staging` branch.

symphorien commented:
I would really like to see this merged in one form or another.

7c6f434c commented on pkgs/os-specific/linux/audit/audit-strndupa-rawmemchr-compat.patch:
```diff
@@ -0,0 +1,141 @@
+From d579a08bb1cde71f939c13ac6b2261052ae9f77e Mon Sep 17 00:00:00 2001
+From: Steve Grubb <sgrubb@redhat.com>
+Date: Tue, 26 Feb 2019 18:33:33 -0500
+Subject: [PATCH] Add substitue functions for strndupa & rawmemchr
```
```suggestion
Subject: [PATCH] Add substitute functions for strndupa & rawmemchr
```

wfranzini commented:
Later today, I'm going to open a new PR starting from staging.

wfranzini commented:
I'll continue working on the cross compilation to musl32 in PR #89471
